src/app/clusters/zone-management-server/zone-util.h

```diff
@@ -0,0 +1,140 @@
+/*
+ * Utility Class providing functionality around various checks required for
+ * zones, such as self-intersection, duplicates, etc.
+ *
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <vector>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ZoneManagement {
+
+class ZoneUtil
+{
+public:
+    using TwoDCartesianVertexStruct = Structs::TwoDCartesianVertexStruct::Type;
+
+    static bool AreTwoDCartVerticesEqual(const TwoDCartesianVertexStruct & v1, const TwoDCartesianVertexStruct & v2)
+    {
+        return v1.x == v2.x && v1.y == v2.y;
+    }
+
+    // Method to check for self-intersecting zones
+    static bool IsZoneSelfIntersecting(const std::vector<TwoDCartesianVertexStruct> & vertices)
+    {
+        size_t vertexCount = vertices.size();
+
+        // Iterate through all pairs of non-adjacent segments
+        for (size_t i = 0; i < vertexCount; ++i)
+        {
+            TwoDCartesianVertexStruct p1 = vertices[i];
+            TwoDCartesianVertexStruct q1 = vertices[(i + 1) % vertexCount];
+
+            for (size_t j = i + 1; j < vertexCount; ++j)
+            {
+                TwoDCartesianVertexStruct p2 = vertices[j];
+                TwoDCartesianVertexStruct q2 = vertices[(j + 1) % vertexCount];
+
+                // Skip segments that share an endpoint (i.e., adjacent segments)
+                bool segments_are_adjacent = false;
+                if (AreTwoDCartVerticesEqual(p1, p2) || AreTwoDCartVerticesEqual(p1, q2) || AreTwoDCartVerticesEqual(q1, p2) ||
+                    AreTwoDCartVerticesEqual(q1, q2))
+                {
+                    segments_are_adjacent = true;
+                }
+
+                // Also a logical adjacency check for sequence
```

Comment:
Why?  If j = i+1, then p2 == q1, and why are we doing this extra check?

At the very least, this needs to document why such a check should be done.